### PR TITLE
LUN-288: Adjust exercise finished screen

### PR DESCRIPTION
**Pull request by SteffiStuffel**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 1]
```diff
@@ -56,66 +67,103 @@ interface Props {
...
+      setMessage(labels.results.feedbackBad)
+      setResultColor(theme.colors.incorrect)
+      setButtonText(labels.results.button.repeat)
+      setNavigationAction(() => () => repeatExercise())
+      setIcon(<SadSmileyIcon width={iconSize} height={iconSize} />)
+    }
+  }, [results])
+
   return (
     <Root>
```
Comment (f1sh1918) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:136]: the `root` container has to be wrapped with a SafeAreaView to get the close button clickable for ios
```js
    <SafeAreaView style={{flex:1}}>
```
probably we have to wrap every screen with that container to ensure working all the ui elements on ios devices with notches. I can create a ticket for that. It's important not to wrap the navigation stuff just fyi

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 2]
```diff
@@ -56,66 +67,103 @@ interface Props {
...
-      setMessage(labels.results.feedbackMedium)
-    } else {
-      setMessage(labels.results.feedbackBad)
-    }
-  }, [results])
+  const [message, setMessage] = useState<string>('')
+  const [resultColor, setResultColor] = useState<Color>(theme.colors.background)
+  const [buttonText, setButtonText] = useState<string>('')
+  const [navigationAction, setNavigationAction] = useState<() => void>(() => undefined)
+  const [icon, setIcon] = useState<ReactElement>()
```
Comment (f1sh1918) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:77]:
```suggestion
  const [icon, setIcon] = useState<SVGElement | null>(null)
```

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 3]
```diff
@@ -56,66 +67,103 @@ interface Props {
...
+        disciplineTitle,
+        closeExerciseAction
+      })
+    }
   }
 
-  const checkResults = (): void => {
-    navigation.navigate('Result', route.params)
+  const navigateToNextModule = (): void => {
+    navigation.pop(2)
```
Comment (f1sh1918) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:99]: this navigates me to the home screen when i finished the last exercise of a module. Why not directly navigate to this screen but pop back?

**Review by f1sh1918 — COMMENTED** [review]
Really nice work :) Just the SafeArea View is needed to be able to close the ExerciseScreen on ios.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx (hunk 3 above)
Comment (SteffiStuffel) [reply src/routes/exercise-finished/ExerciseFinishedScreen.tsx:99]: It is not trivial to navigate to the next module or even the parent discipline selection. This solution is kind of a quick fix. Have a look at https://issues.tuerantuer.org/browse/LUN-364

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx (hunk 1 above)
Comment (SteffiStuffel) [reply src/routes/exercise-finished/ExerciseFinishedScreen.tsx:136]: Yes please open a ticket. Will fix this screen now, nevertheless.

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
Tested on android, nice work! I like the design :)

**Review by f1sh1918 — APPROVED** [review]
works fine now on ios

File: assets/images/index.ts [hunk 4]
```diff
@@ -42,8 +43,10 @@ import LockIcon from './lock-icon.svg'
 import LunesIcon from './lunes-icon.svg'
 import MenuIcon from './menu-icon.svg'
 import OpenLockIcon from './open-lock-icon.svg'
+import PartyHornIcon from './party-horn-circle-icon-white.svg'
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply assets/images/index.ts:46]: Used ParyHorn as this is the font-awesome naming.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 5]
```diff
@@ -56,69 +68,109 @@ interface Props {
...
+        setIcon(<PartyHornIcon width={iconSize} height={iconSize} />)
+      }
+    } else {
+      setMessage(labels.results.feedbackBad)
+      setResultColor(theme.colors.incorrect)
+      setButtonText(labels.results.button.repeat)
+      setNavigationAction(() => () => repeatExercise())
+      setIcon(<SadSmileyIcon width={iconSize} height={iconSize} />)
+    }
+  }, [results])
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/exercise-finished/ExerciseFinishedScreen.tsx:134]: Fixed it here and opened a ticket for general adjusting this.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 6]
```diff
@@ -56,69 +68,109 @@ interface Props {
...
-      setMessage(labels.results.feedbackMedium)
-    } else {
-      setMessage(labels.results.feedbackBad)
-    }
-  }, [results])
+  const [message, setMessage] = useState<string>('')
+  const [resultColor, setResultColor] = useState<Color>(theme.colors.background)
+  const [buttonText, setButtonText] = useState<string>('')
+  const [navigationAction, setNavigationAction] = useState<() => void>(() => undefined)
+  const [icon, setIcon] = useState<SVGElement | null>(null)
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/exercise-finished/ExerciseFinishedScreen.tsx:78]: Really good idea. Thanks.

File: src/constants/labels.json [hunk 7]
```diff
@@ -93,19 +93,21 @@
...
     "feedbackBad": "Nicht aufgeben! \nVersuche es noch einmal!",
-    "checkEntries": "Eingaben überprüfen",
-    "retryExercise": "Übung wiederholen",
-    "unlockedExercise": "Glückwunsch, deine nächste Übung ist jetzt freigeschaltet.",
-    "wrong": "falsche",
-    "similar": "ähnliche",
-    "correct": "richtige",
+    "finishedModule": "Super, du hast das Modul abgeschlossen",
+    "unlockExercise": {
+      "part1": "Glückwunsch",
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/constants/labels.json:100]: Added comma to ExerciseFinishedScreen.

File: src/routes/exercise-finished/__tests__/ExerciseFinishedScreen.spex.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      unlockedNextExercise
+    }
+  })
+
+  it('should render and handle button click for unlocked next exercise', () => {
+    const route = getRoute(1, true, true)
+    const { getByText } = render(<ExerciseFinishedScreen route={route} navigation={navigation} />)
+    expect(
+      getByText(`${labels.results.unlockExercise.part1} ${EXERCISES[2].title} ${labels.results.unlockExercise.part2}`)
+    ).toBeDefined()
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/exercise-finished/__tests__/ExerciseFinishedScreen.spex.tsx]: I think using the labels makes more sense, as it is less like snapshot testing and we don't have labels that are widely spread over the code. But sometime I also think our tests just "imitating" the code.

**Review by steffenkleinle — APPROVED** [review]
Github are you drunk?
image

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx [hunk 9]
```diff
@@ -71,21 +72,18 @@ const ExerciseFinishedScreen = ({ navigation, route }: Props): ReactElement => {
...
-  const [icon, setIcon] = useState<SVGElement | null>(null)
-
-  const repeatExercise = (): void =>
-    navigation.navigate(EXERCISES[exercise].screen, {
-      documents,
-      disciplineId,
-      disciplineTitle,
-      closeExerciseAction
-    })
+  const repeatExercise = useCallback(
```
[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/routes/exercise-finished/ExerciseFinishedScreen.tsx:75]: by mistake, will undo

**Pull request merged**